Sudokumb application.

SurferJeffAtGoogle opened the pull request.

jsimonweb commented on applications/sudokumb/WebApp/Views/Home/Index.cshtml:
```diff
@@ -0,0 +1,155 @@
+            </dt>
+            <dd>
+                Sudokumb stores user information like email addresses and passwords in 
+                <a href="https://cloud.google.com/datastore/">Google Cloud Datastore></a>.
```
Update to remove orphaned ">" symbol after "Datastore"

SurferJeffAtGoogle replied:
Done.

jsimonweb commented on applications/sudokumb/WebApp/Views/Account/AccessDenied.cshtml:
```diff
@@ -0,0 +1,24 @@
+}
+
+<header>
+    <h2 class="text-danger">ViewData["Title"]</h2>
```
Add missing "@" prefix symbol before ViewData["Title"]

SurferJeffAtGoogle replied:
Done.

jsimonweb commented on applications/sudokumb/README.md:
```diff
@@ -0,0 +1,142 @@
+
+1.  Open ***Sudokumb.sln**.
+
+2.  In the **Solution Explorer** window, right-click the **Sudokumb** project 
```
Suggestion to update this to be:

In the **Solution Explorer** window, right-click the **Sudokumb** solution

SurferJeffAtGoogle replied:
Done.

jsimonweb commented on applications/sudokumb/README.md:
```diff
@@ -0,0 +1,142 @@
+3.  Click **Multiple startup projects:**.  Change the **Action** from **None**
+    to **Start** for the projects **WebApp** and **WebSolver**.
+
+4.  Click **OK** to close the **Solution 'Sudokumb' Propert Pages** window.
```
Update to be:

Click **OK** to close the **Solution 'Sudokumb' Property Pages** window.

SurferJeffAtGoogle replied:
Done.

simonz130 commented on applications/sudokumb/KmsDataProtectionProvider/KmsDataProtectionProvider.cs:
```diff
@@ -0,0 +1,236 @@
+        /// </summary>
+        public string ProjectId { get; set; }
+        /// <summary>
+        /// global, us-east1, etc.
```
worth pointing to the link where all locations are defined

SurferJeffAtGoogle replied:
Done.

SurferJeffAtGoogle commented on applications/sudokumb/Add-KmsPermissionsToAppEngine.ps1:
```diff
@@ -0,0 +1,30 @@
+# License for the specific language governing permissions and limitations under
+# the License.
+
+$accounts = gcloud iam service-accounts list --format=json | ConvertFrom-Json
```
Done.

SurferJeffAtGoogle commented on applications/sudokumb/AppEngine.md:
```diff
@@ -0,0 +1,9 @@
+### <a name="appengine"><img src="http://cloud.google.com/_static/images/cloud/products/logos/svg/appengine.svg" width=64> Sudokumb automatically scales on [App Engine](https://cloud.google.com/appengine/docs/flexible/dotnet/).</a>
+
+Solving a Sudoku puzzle the dumb way is an intentinally CPU and network intensive task.
```
Done.

SurferJeffAtGoogle commented on applications/sudokumb/Counters/Counters.cs:
```diff
@@ -0,0 +1,128 @@
+    public class LockingCounter : ICounter
+    {
+        long _count = 0;
+        readonly object _thisLock = new object();
```
Done.

SurferJeffAtGoogle commented on applications/sudokumb/Counters/Counters.cs:
```diff
@@ -0,0 +1,128 @@
+        long Count { get; }
+    }
+
+    public class UnsynchronizedCounter : ICounter
```
Done.

SurferJeffAtGoogle commented on applications/sudokumb/Counters/Counters.cs:
```diff
@@ -0,0 +1,128 @@
+        {
+            get
+            {
+                lock (_thisLock)
```
In the benchmark, the get Count operation happens only 3 times per second, while Increase() is called 10,000 to 100,000 times per second, so I don't think it will affect performance of the benchmark.  I'd rather stick with the simpler code.

SurferJeffAtGoogle commented on applications/sudokumb/DatastoreUserStore/DatastoreUserStore.cs:
```diff
@@ -0,0 +1,340 @@
+            USER_KEY = "key";
+
+        public DatastoreUserStore(DatastoreDb datastore,
+            ILogger<DatastoreUserStore<U>> logger)
```
I haven't seen the pattern you describe in any code I've come across.

And, the code generated by "dotnet new mvc" uses the same pattern I use here.  For example, look at AccountController's constructor:
```cs
  public AccountController(
            UserManager<ApplicationUser> userManager,
            SignInManager<ApplicationUser> signInManager,
            IEmailSender emailSender,
            ILogger<AccountController> logger,
            IOptions<AccountOptions> options,
            IUserRoleStore<ApplicationUser> userRoleStore)
        {
```

SurferJeffAtGoogle commented on applications/sudokumb/DatastoreUserStore/DatastoreUserStore.cs:
```diff
@@ -0,0 +1,340 @@
+            user.UserName = userName;
+            return Task.CompletedTask;
+        }
+        void IDisposable.Dispose()
```
There's nothing to dispose, and there is no parent class.  This class inherits from an *interface* with a `Dispose` method.

SurferJeffAtGoogle commented on applications/sudokumb/GameLibTest/GameBoardTest.cs:
```diff
@@ -0,0 +1,107 @@
+
+namespace Sudokumb
+{
+    public class Tests
```
Done.

SurferJeffAtGoogle commented on applications/sudokumb/GameLibTest/GameBoardTest.cs:
```diff
@@ -0,0 +1,107 @@
+            "3  |   |1  ";
+        readonly
+
+                GameBoard _boardA = GameBoard.Create(s_boardA);
```
Done.

SurferJeffAtGoogle commented on applications/sudokumb/GameLib/GameBoard.cs:
```diff
@@ -0,0 +1,293 @@
+
+        private static string GetRow(int rowNumber, string board)
+        {
+            Debug.Assert(rowNumber >= 0 && rowNumber < 9);
```
Done.

SurferJeffAtGoogle commented on applications/sudokumb/GameLib/GameBoard.cs:
```diff
@@ -0,0 +1,293 @@
+
+        public static GameBoard Create(string board) => new GameBoard()
+        {
+            Board = new string(board.Where((c) =>
```
Done.

SurferJeffAtGoogle commented on applications/sudokumb/GameLib/GameBoard.cs:
```diff
@@ -0,0 +1,293 @@
+            set
+            {
+                // Validate the board.
+                if (value.Length != 81)
```
Done.

SurferJeffAtGoogle commented on applications/sudokumb/DatastoreUserStore/DatastoreUserStore.cs:
```diff
@@ -0,0 +1,340 @@
+            }
+            catch (Grpc.Core.RpcException e)
+            {
+                return IdentityResult.Failed(new IdentityError()
```
Done.

SurferJeffAtGoogle commented on applications/sudokumb/DatastoreUserStore/DatastoreUserStore.cs:
```diff
@@ -0,0 +1,340 @@
+        public async Task<IdentityResult> CreateAsync(U user,
+            CancellationToken cancellationToken)
+        {
+            _logger.LogDebug("CreateAsync({0})", user.NormalizedEmail);
```
Done.

SurferJeffAtGoogle commented on applications/sudokumb/DatastoreUserStore/DatastoreRoleStore.cs:
```diff
@@ -0,0 +1,138 @@
+
+        Key KeyFromRoleId(string roleId) => _roleKeyFactory.CreateKey(roleId);
+
+        Entity RoleToEntity(R role)
```
I don't see a benefit to extracting to a separate class.  All the calls to these functions are within this class.

SurferJeffAtGoogle commented on applications/sudokumb/GameLibTest/GameLibTest.csproj:
```diff
@@ -0,0 +1,24 @@
+  </ItemGroup>
+
+<ItemGroup>
+  <None Update="SampleBoard.txt" CopyToOutputDirectory="PreserveNewest" />
```
What does this mean?

SurferJeffAtGoogle commented on applications/sudokumb/GameLib/GameBoard.cs:
```diff
@@ -0,0 +1,293 @@
+        }
+    }
+
+    public class BadGameBoardException : Exception
```
I read the article a couple times, and I don't see the benefit of implementing this exception the way thee article describes.  I see it would take many more lines of code that are never invoked, and they would need to be maintained.

SurferJeffAtGoogle commented:
> Hi Jeff, it's a long change list. I looked up to KmsDataProtectionProvider so far. Will look at the rest next week.

Thank you Simon and Jonathan.  The code and READMEs look much better already, thanks to your comments.  I have addressed all comments to date.
